README
Redis::JobQueue package version 1.18
====================================

The Redis::JobQueue package is a set of Perl modules which provides a
simple job queue with Redis server capabilities.

The Redis::JobQueue package was written, tested, and found working on
recent Linux distributions.

INSTALLATION

To install this package type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

In order to install and use this package you will need Perl version
5.010 or better. Some modules within this package depend on other
packages that are distributed separately from Perl. We recommend that
you have the following packages installed before you install
Redis::JobQueue :

   Data::UUID
   Digest::SHA1
   List::MoreUtils
   Mouse
   Params::Util
   Redis
   Storable
   Try::Tiny

The Redis::JobQueue package has the following optional dependencies:

   Net::EmptyPort
   Test::Deep
   Test::Exception
   Test::NoWarnings
   Test::RedisServer

If the optional modules are missing, some "prereq" tests are skipped.

The installation of the missing dependencies can either be accomplished
through your OS package manager or through CPAN (or downloading the source
for all dependencies and compiling them manually).

TESTING

You need to have a running Redis instance before launching the test.

DOCUMENTATION

Every Redis::JobQueue package module has a man page.
A simple description of each module is included below in the SEE ALSO section.

SOURCE CODE
Redis::JobQueue is hosted on GitHub:
https://github.com/TrackingSoft/Redis-JobQueue

SEE ALSO

The basic operation of the Redis::JobQueue package modules:

   Redis::JobQueue      - Object interface for creating and executing jobs
                          queues, as well as monitoring the status and results
                          of jobs
   Redis::JobQueue::Job - Object interface for creating and manipulating jobs
   Redis::JobQueue::Util - String manipulation utilities

COPYRIGHT AND LICENCE

Copyright (C) 2012-2016 by TrackingSoft LLC.

This package is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See perlartistic at
http://dev.perl.org/licenses/artistic.html .

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.


Changes
Revision history for Redis::JobQueue package

    CHANGES:
    - Ability to force re-connection with Redis on error ('reconnect_on_error' attribute).
    - $DEFAULT_CONNECTION_TIMEOUT, $DEFAULT_OPERATION_TIMEOUT constants added.
    - connection_timeout, operation_timeout attributes added.
    - Added the service module Redis::JobQueue::Util.
    - Attempt to protect against used_memory > maxmemory.
    - Added the 'queue_length' method.

1.18 Thu May  31 2016
    BUGFIXES:
    - Fix MANIFEST.

1.17 Thu May  26 2016
    BUGFIXES:
    - Tests corrected.
    - Require perl 5.10 core modules version.
    - Data::Dumper 'Sparseseen' method not used now.

1.16 Thu Sep  03 2015
    CHANGES:
    - In the parameters of the constructor 'redis' may be either a Redis object
      or a server string or a hash reference of parameters to create a Redis object.
    - Redis::JobQueue does not close connection to redis server (including '->quit' method)
      if it was provided by user.

1.15 Tue Oct  28 2014
    CHANGES:
    - Time::HiRes::time is used for storing internal attributes.
    - Redis::JobQueue->queue_status returns a floating seconds since the epoch
      for the following information related to the queue:
      max_job_age, min_job_age, lifetime.
    - The Redis::JobQueue->quit method does not close connection to Redis during
      global destruction phase (Perl version 5.14 and above).
    FIXES:
    - Fixed test 't/01_JobQueue/16_utf8.t'

1.13 Tue May  27 2014
    CHANGES:
    - Re-factoring of tests.

1.12 Thu May  23 2014
    CHANGES:
    - Added argument 'check_maxmemory' to the Redis::JobQueue->new
      (thanks to Danny Sterne).

1.11 Fri Mar  14 2014
    CHANGES:
    - Change the port used for testing.

1.10 Thu Feb  20 2014
    CHANGES:
    - Small test improvement.

1.09 Fri Jan  10 2014
    CHANGES:
    - Added method Redis::JobQueue->get_next_job_id.

1.08 Fri Nov  15 2013
    CHANGES:
    - Deleted argument 'encoding' from the Redis::JobQueue->new
      (Starting with version 1.963, Redis module doesn't support encoding).

1.07 Mon Oct  28 2013
    CHANGES:
    - Added to META.yml the list of the modules provided by this distribution.

1.06  Fri Oct  25 2013
    CHANGES:
    - 'delete_job' removes job id record from the queue list.
      This helps to properly diagnose cases when job data was removed by Redis OOM handler
      rather than explicitly deleted by user.
    - Use single hash for storing all job data on Redis server.
    - Removed 'NS_METADATA_SUFFIX' constant.
    FIXES:
    - Fix runtime loading of Test::*

1.05  Wed Sep  18 2013
    CHANGES:
    - Use Try::Tiny for proper preservation of $@ .

1.04  Thu Sep  05 2013
    CHANGES:
    - Method 'clear_variability' renamed to 'clear_modified'.
    - Redis server updates only updated data, if you change the metadata,
      not all of the metadata.
    - Change the 'completed' attribute sets 'failed' to 0,
      the change 'failed' sets 'completed' to 0.
      The attributes 'completed' and 'failed' are mutually exclusive.
    - Method 'update_job' accesses the Redis server only if there is some
      data to update.
    FIXES:
    - $@ is localized before we attempt to catch an exception.

1.03  Tue Jun  25 2013
    FIXES:
    - Added description of the Redis::JobQueue::Job attributes:
      'progress', 'message'.

1.02  Mon Jun  17 2013
    CHANGES:
    - Added standard job field: Time when job was 'failed' (user settable).
    - Added a method 'last_error'.

    FIXES:
    - 'add_job', 'update_job', 'load_job' not cause 'clear_variability' for metadata.

1.01  Tue May  28 2013
    CHANGES:
    - Minor changes in documentation.

    FIXES:
    - Checking OS in Makefile.PL
      Tests hang on Windows.

1.00  Mon May  07 2013

    !!! This is a major upgrade that changes API. It's strongly advised to study
    revised documentation and update your code where necessary !!!

    CHANGES:
    - Needs Redis server version 2.6 or higher as module uses Redis Lua scripting.
    - Improved handling of transactions.
    - Method Redis::JobQueue::Job->job_attributes returns a sorted list.
    - Fields 'result' and 'workload':
      It is possible to use any data structure (SCALAR, HASH, ARRAY, OBJECT).
      It is automatically serialized.
      Allowable size of the data is determined based on the size of the
      serialized string.
    - Added standard job fields:
      Times when job was: 'created', 'started', 'completed', 'updated'
      (user settable),
      Job 'progress' (0...1) and arbitrary status 'message'.
      Appropriate methods added.
    - The job 'updated' field automatically changes with every change to a job.
    - Added a method 'Redis::JobQueue::Job->elapsed'.
    - A set of constants to reflect job status:
      CREATED, WORKING, FAILED, COMPLETED, the user can specify own value
      for status.
      Moved from module Redis::JobQueue to module Redis::JobQueue::Job.
    - Added working with the jobs metadata
      (additional information related to the job):
      It is automatically serialized.
      Methods 'Redis::JobQueue::Job->meta_data',
      'Redis::JobQueue->get_job_meta_fields'.
    - Constant MAX_DATASIZE moved to module Redis::JobQueue.
    - No longer used constants EMAXMEMORYPOLICY, STATUS_DELETED, EXPIRE_DELETED.
    - Added a method 'queue_status' for monitoring the state of the queue.
    - Method 'get_job_status' removed.
      Method 'get_job_data' added instead for receiving data from the standard
      set of fields.
    - Method 'delete_job' removes Redis data structures that are associated
      with the job.
    - Method 'get_jobs' renamed to 'get_job_ids'.
      Jobs can be filtered by queue name or status.
    - Method 'add_job' updates Redis::JobQueue::Job object, if it was passed
      as the first argument.
    - 'encoding => undef' for Redis used, if 'new' invoked without
      the first argument being an object of 'Redis::JobQueue' or 'Redis' class.

Makefile.PL
use 5.010;
use strict;
use warnings;

if ( $^O eq 'MSWin32' ) {
    print STDERR "OS unsupported: $^O\n";
    exit( 255 );
}

use File::Basename;
use ExtUtils::MakeMaker qw( prompt );

my %config = (
    NAME                        => 'Redis::JobQueue',
    VERSION_FROM                => 'lib/Redis/JobQueue.pm',
    MIN_PERL_VERSION            => '5.010',
# more to the list $Module::CoreList::version{5.010}
    PREREQ_PM                   => {
        'Data::UUID'            => 1.217,
        'Digest::SHA1'          => 2.13,
        'List::MoreUtils'       => 0.33,
        'Mouse'                 => 0,
        'Params::Util'          => 1.07,
        'Redis'                 => 1.976,
        'Storable'              => 2.18,
        'Try::Tiny'             => 0.18,
        },
    TEST_REQUIRES               => {
        'Test::Deep'            => 0.109,
        'Test::Exception'       => 0.31,
        'Test::NoWarnings'      => 1.04,
        'Test::RedisServer'     => 0.04,
        'Test::TCP'             => 2.11,
        },
    META_MERGE          => {
        no_index                => {
            directory               => [
#                't',
#                'inc',
                'xt',
                'ex',
            ]
        },
        provides                => {
            'Redis::JobQueue'       => {
                file                    => 'lib/Redis/JobQueue.pm',
                version                 => '1.18'
            },
            'Redis::JobQueue::Job'  => {
                file                    => 'lib/Redis/JobQueue/Job.pm',
                version                 => '1.18'
            },
            'Redis::JobQueue::Util' => {
                file                    => 'lib/Redis/JobQueue/Util.pm',
                version                 => '1.18'
            },
        },
        'meta-spec'             => { version => 2 },
        resources               => {
            repository              => {
                type                    => 'git',
                url                     => 'git://github.com/TrackingSoft/Redis-JobQueue.git',
                web                     => 'http://github.com/TrackingSoft/Redis-JobQueue',
            },
        },
    },
    ( $] >= 5.005 ?
        (
        ABSTRACT        => 'Redis::JobQueue simple job queue with Redis server capabilities.',
        AUTHOR          => 'TrackingSoft LLC <sgladkov@trackingsoft.com>'
        ) : ()
    ),
    dist        => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean       => { FILES => "Debian_CPANTS.txt Redis-*.tar.gz" },
    LICENSE     => 'perl',
#    NO_MYMETA   => 1,
);

my @patterns = qw( t/*.t t/*/*.t );
$config{test} = { TESTS => join ' ', map { glob } @patterns };

if( $ENV{AUTHOR_TESTS} ) {
    @patterns = qw( xt/author/*.t xt/author/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

if( $ENV{RELEASE_TESTS} ) {
    @patterns = qw( xt/release/*.t xt/release/*/*.t );
    $config{test}->{TESTS} .= ' '.join ' ', map { glob } @patterns;
}

mkdir 'inc' unless -d 'inc';

ExtUtils::MakeMaker::WriteMakefile( %config );
